### PR TITLE
Working Group Proposal for Reference Types

### PR DESCRIPTION
**Pull request by lachie83**
This is a culmination of many people's work over the last few months to introduce reference types to OCI as part of the [OCI Reference Types WG](https://github.com/opencontainers/tob/blob/main/proposals/wg-reference-types.md).

This PR contains the changes from our chosen proposal, [Proposal E](https://github.com/opencontainers/wg-reference-types/blob/main/docs/proposals/PROPOSAL_E.md), relevant to the image-spec. We're looking forward to hearing feedback from image-spec maintainers on these changes!

Associated distribution-spec PR - https://github.com/opencontainers/distribution-spec/pull/335

### CODE REVIEW

**Comment by vbatts** [comment]
OHMAN

File: artifact.md
Comment (mounchin) [inline artifact.md:53]: As the spec allows user defined annotations, do we want to cap the number of annotations stored or returned as part of `referrers` API ?

Comment (sajayantony) [reply artifact.md:53]: The current proposal is that all annotations would be returned. If we were to trim down the list, then we need define the criteria.  For example, only the first 10 annotations would be returned.  That still doesn't cap the size of the payload since a single key could actually be large enough. 

I think it would be good to raise this issue in the distribution spec since the referrers response payload and registry implementations needs to now ensure that data from all manifests are returned.  https://github.com/opencontainers/distribution-spec/pull/335 /cc @michaelb990

Comment (sudo-bmitch) [reply artifact.md:53]: The index we return is rather minimal, so having enough annotations to cause this to exceed the recommended 4MB limit is likely to also exceed that limit in the referrer's manifest too. I think any limits we impose should be set on the original manifest, and not this API, to minimize complexity.

Comment (mounchin) [reply artifact.md:53]: Lets say worst case, there are like 20 artifacts in the referrers API reponse and each had say about ~3MB of annotations, then the API reponse payload would be very large. In these cases what should the API return? May be capping the reponse size per artifact?

**Comment by sudo-bmitch** [comment]
@opencontainers/image-spec-maintainers given the size of this, we are giving this 2-4 weeks for any objections before merging. It was presented on the OCI call today and hopefully a recording will be posted soon. After merging it still won't be a tagged release, so if there are any issues raised, we can always open another PR.

**Comment by lachie83** [comment]
We would kindly like to ask for reviews and/or approval from maintainers by August 25th, 2022. Please let us know if you don't think this is attainable.

**Comment by lachie83** [comment]
Adding link to similar comment on distribution-spec PR. https://github.com/opencontainers/distribution-spec/pull/335#issuecomment-1212296424

File: artifact.md
Comment (sudo-bmitch) [reply artifact.md:53]: At that point, the response would trigger pagination, with one descriptor per page. It also wouldn't be supported to include more than one of those descriptors in the tag schema fallback. My hope is that manifests like that result in a bad enough UX, that bug reports get opened on the source and alternatives are used.

The downside of capping what is returned in the API response, is that any partial descriptor would then need to be pulled as a full manifest to determine if the client side filters match (the filtered annotation could be one that was trimmed). We would also need to provide an indicate that the descriptor is a partial response so clients know when they need to fall back to pulling the full manifest directly. When thinking of client implementations, that would probably be one pull for the API, one of the manifest by the artifact listing method, and then another pull of the manifest by the artifact get method, unless the manifest is cached between those methods.

File: descriptor.md
Comment (shizhMSFT) [inline descriptor.md:59]: The `artifactType` is defined twice. Should we merge both or drop one?

Comment (lachie83) [reply descriptor.md:59]: Opened https://github.com/oci-playground/image-spec/pull/14

File: manifest.md
Comment (mikebrow) [inline manifest.md:70]: can we clarify if there is a restriction or not for media type of the refers descriptor.. image manifest and image index manifest for example..

**Review by mikebrow — COMMENTED** [review]
can we, should we add something over in image-index.md, such as another example, for using the image.index manifest to hold artifact manifests in the list of manifests? See example provided in distribution api..

File: artifact.md
Comment (mikebrow) [inline artifact.md:20]: the MAY language, here, doesn't sync with the "is" and MUST language in other spots..

might be useful to make media-types.md media-and-artifact-types and put a paragraph there talking about any MUST/MAY language wrt artifact types?

File: artifact.md
Comment (wmesard) [inline artifact.md:25]: I have a use case where each element in the `blobs` array is associated with the corresponding element in the `layers` array of the referent OCI image. In other words, the length of the two arrays are always equal. However, a blob in the referring artifact may be omitted for any particular image layer.

This can be addressed by adding to this line: "Any element in the list may also be `null`."

Posting here for comments. I can open a new PR if that's easier.

Comment (sudo-bmitch) [reply artifact.md:25]: I suspect a null descriptor entry may cause issues for tools. What if an annotation in the descriptor was used to indicate which layer it is referencing. That could be a layer digest, or just a list index number.

File: manifest.md
Comment (sudo-bmitch) [reply manifest.md:70]: We shouldn't have a restriction here. An artifact shipped with an Artifact manifest or this Image manifest can refer to any other manifest, including an Index. (I.e. You can attach a signature on a multi-platform manifest.)

Comment (mikebrow) [reply manifest.md:70]: before suggesting text changes here to include that clarification..  Do we have to worry about/restrict wider dependency loops here outside the obvious don't refer to yourself loop? E.g. an image manifest with a refers to it's own image.index.. an image manifest with a refers to an artifact that refers to this image manifest...

Comment (sudo-bmitch) [reply manifest.md:70]: For loops, I'm looking at whether the refers descriptor can point to a normal child of the manifest. If all children are blobs (as with Image and Artifact manifests), and not manifests (like we have with an Index), you can't point to a manifest that's a child.

It's a reverse pointer, so we aren't worried about parents (both pointers would be going the same direction). And you can't know the digest of the parent in advance without a hash collision/prediction.

File: artifact.md
Comment (wmesard) [reply artifact.md:25]: This PR is defining the spec that the tools will have to conform to, no? But okay; I can go with the annotation approach. Thanks.

File: manifest.md
Comment (mikebrow) [reply manifest.md:70]: there may be an opportunity in platform image index.. where an artifact manifest included in the platform image index could be to an artifact that does not have a refers.. and in that case the artifact could be implied to be for that index (refer) by inclusion... thus allowing an index to have an artifact that refers to itself without incurring the loop restriction.. thinking out loud.

File: artifact.md
Comment (sudo-bmitch) [inline artifact.md:23]: We need to change "string" to an array of descriptors.

File: artifact.md
Comment (fawadasaurus) [inline artifact.md:17]: Would it be possible to allow `config` here as `OPTIONAL`?  In my use case, I want to use an OCI artifact for packaging an application and want to use the `config` to provide execution details for the application.

Comment (sudo-bmitch) [reply artifact.md:17]: You can push an additional blob with a distinct media type. Or you can use the image manifest with a custom config. I don't believe anything we are doing requires an artifact to use the artifact manifest. The existing OCI Artifact guidance using the image manifest is still valid, and I'd even recommend it since it will take time for registries to support the new artifact manifest type.

Comment (fawadasaurus) [reply artifact.md:17]: > you can use the image manifest with a custom config

Yes.  Thats exactly what I am doing now, but I was hoping that when the artifact spec is released, that I would not be having to have a technically non-standard image.  My application does not use layers so the image spec with `layers` is not technically valid.

I understand that the spec will take some time to be supported but I would like that once the time comes, I am able to use the OCI distribution without breaking the `rules`:

Here is the spec description for `layers`
> The array MUST have the base layer at index 0. Subsequent layers MUST then follow in stack order (i.e. from layers[0] to layers[len(layers)-1]). The final filesystem layout MUST match the result of [applying](https://github.com/opencontainers/image-spec/blob/main/layer.md#applying-changesets) the layers to an empty directory. The [ownership, mode, and other attributes](https://github.com/opencontainers/image-spec/blob/main/layer.md#file-attributes) of the initial empty directory are unspecified.

By adding another blob, it is making it more difficult for existing registry distributions to be used because (for example) harbor already reads the `config` for an image automatically do display in the UI.  So now anyone that wants to use an artifact will need to make their own custom flows that are specific to their own use case to download an individual arbitrary blob from the image as opposed to just downloading a standard manifest config.

Your point is valid that these can be worked around, but I am hoping that the solution is built flexible enough that novel use cases are not `hacks` but can be widely used across all OCI-compliant registry implementations.

Comment (sajayantony) [reply artifact.md:17]: The current goal of the working group was to support attaching references and that was the primary motivation of creating the artifact manifest for the reference type working group without having to hack the image manifest to attach blobs of signatures or sboms. 

The artifact manifest does have potential, and it would be nice to consider if we want to expand its scope to help define other types of images/applications. If we do want to enhance the scope of the artifact manifest I would recommend we can do this once reference type work is merged and discuss adding more fields (like the data field in the descriptor https://github.com/opencontainers/image-spec/pull/826). 

Does that sound reasonable?

Comment (sudo-bmitch) [reply artifact.md:17]: > By adding another blob, it is making it more difficult for existing registry distributions to be used because (for example) harbor already reads the `config` for an image automatically do display in the UI. So now anyone that wants to use an artifact will need to make their own custom flows that are specific to their own use case to download an individual arbitrary blob from the image as opposed to just downloading a standard manifest config.

I haven't pulled up the Harbor UI in a while. What metadata are they extracting from the `config`? One of the goals we had was to move the config mediaType to artifactType, and then any data could be an annotation. I'm wondering if we have all the data needed without introducing another blob.

There's also nothing preventing us from defining a new artifact metadata json schema in the future, with a well known media type, that could be added as a blob to any artifact.

Comment (fawadasaurus) [reply artifact.md:17]: Thanks for the feedback.  I am certainly not wanting to derail any efforts.  I will wait for the spec to get merged and then will post any feedback as an issue and we can address in the future.

> What metadata are they extracting from the config?

Harbor reads the JSON in the config file and displays it directly in the UI:

config.json
```
{
    "flows": [
        "flow.wafl"
    ],
    "permissions": [
        "permissions.wafl"
    ],
    "interfaces": [
        "add.apex",
        "delete.apex",
        "edit.apex",
        "list.apex"
    ]
}
```
image

File: specs-go/v1/descriptor.go
Comment (mikebrow) [inline specs-go/v1/descriptor.go:49]: ?

Comment (sudo-bmitch) [reply specs-go/v1/descriptor.go:49]: This is used in the Referrers listing.

**Review by sudo-bmitch — APPROVED** [review]

File: specs-go/v1/descriptor.go
Comment (mikebrow) [reply specs-go/v1/descriptor.go:49]: if this refers descriptor example is correct, the definition could be improved by saying
```suggestion
        // a refers descriptor use of artifactType SHOULD be to specify the media type of the artifact being referred to
	ArtifactType string `json:"artifactType,omitempty"`
```

**Comment by vbatts** [comment]
looking at the git history, it seems there was a merge done instead of rebase, and now there is a bubble with #919 in it.

File: artifact.md
Comment (vbatts) [inline artifact.md:29]: for clarity, this is _the_ field that "refers to" the originating object?

File: artifact.md
Comment (vbatts) [inline artifact.md:3]: nit, each sentence to it's own line

File: descriptor.md
Comment (vbatts) [inline descriptor.md:3]: Is this a space to clarify that descriptors are not exclusive to "An OCI  image"?

File: descriptor.md
Comment (vbatts) [inline descriptor.md:80]: I wonder if github markdown handles ebnf instead of just text here (and below)?

File: specs-go/v1/manifest.go
Comment (vbatts) [inline specs-go/v1/manifest.go:34]: particular reason for using a pointer here? Looks like we're using a mix, and usually there was reasoning for choosing one way or the other

**Review by vbatts — DISMISSED** [review]
This looks good. Nothing breaking and will provide clarity on a tremendous pattern for clients and platforms to provide.

There are a couple nits that need to be cleaned up, but they can be follow-up PRs as they don't affect behavior/functionality.

One unfortunate aspect of merging this from the oci-playground branch is that a number of the commits are noisy as they get shadowed or cleaned up later... That isn't ideal and the _only_ benefit I see in leaving it as is would be so that all the authors preserve their involvement in the commit history.

File: specs-go/v1/descriptor.go
Comment (mikebrow) [reply specs-go/v1/descriptor.go:49]: needs some explanation regarding the parts of the spec that are descriptors and what the artifact type field currently means in those contexts.. will try to draft something probably needs more examples

File: artifact.md
Comment (sudo-bmitch) [reply artifact.md:29]: Yes. The phrasing there can use work. I think it has bounced around several editors until getting punted to "read the distribution-spec".

File: descriptor.md
Comment (sudo-bmitch) [reply descriptor.md:3]: That would make a good follow up PR. This line is showing up in the change log because I switched the list bullets from `*` to `-` to get my markdown linter to stop screaming at the rest of the file. :)

File: specs-go/v1/manifest.go
Comment (sudo-bmitch) [reply specs-go/v1/manifest.go:34]: To support `omitempty`. A struct is never "empty" to Go because it can't be compared to a zero value.

File: descriptor.md
Comment (sudo-bmitch) [reply descriptor.md]: Thanks, I added this.

File: artifact.md
Comment (sudo-bmitch) [reply artifact.md]: Thanks, fixed.

File: artifact.md
Comment (lachie83) [inline artifact.md:33]: cc @vbatts @sudo-bmitch - Does this provide more clarity in response to https://github.com/opencontainers/image-spec/pull/934/files/5240cdaf5cfb2f62f1b6b8860e5561332437270e#r955050333

```suggestion
  This OPTIONAL property specifies a [descriptor](descriptor.md) of another manifest. If defined, the artifact `refers` to, or has a reference to the specified manifest
```

Comment (vbatts) [reply artifact.md:33]: I think so. (please put that sentence on its own newline)

File: descriptor.md
Comment (vbatts) [reply descriptor.md:3]: yeah fair

**Comment by vbatts** [comment]
during discussion on the call, this has enough features and fields that it merits a point release, but not a major version change.

**Review by sudo-bmitch — DISMISSED** [review]

**Review by sajayantony — APPROVED** [review]

**Comment by estesp** [comment]
Thanks to all for the work on this in the WG and getting all this together (and dealing with all the feedback) and ready! IANAM but LGTM! 🎉

**Review by sudo-bmitch — APPROVED** [review]

**Review by sajayantony — APPROVED** [review]

**Pull request merged**
